responsively-org/responsively-app
# added feat: bookmark add, edit, remove, view, save

# ✨ Pull Request
Add Bookmark Feature: Ability to Add, Remove, and View Bookmarks
---
### 📓 Referenced Issue
Fixes: #916
---

### ℹ️ About the PR
I have added the bookmark feature:
#### Bookmark Management
1. Adding Bookmarks:
- Users can add a bookmark by clicking on the star icon on the address bar.
- A flyout appears allowing users to edit the URL and name of the bookmark.
- Users can save the bookmark after making modifications.
2. Removing Bookmarks:
- Users can remove a bookmark using the same method.
- Clicking on the star icon on a bookmark opens the flyout.
- From the flyout, users can remove the bookmark.
3. Viewing Bookmarks:
- To view all bookmarks, users can click on the menu flyout button (vertical three dots) in the top-right corner.
- Clicking on the "Bookmarks" option in the dropdown menu displays a flyout with the list of bookmarks.
4. Empty Bookmarks:
- If no bookmarks are present, the UI text will indicate that there are no bookmarks.
- Users will be informed when there are no bookmarks available to display.

#### Editing Bookmarks
1. Modifying Bookmark Name and URL:
- Users can edit the bookmark's name and URL by clicking on the edit button.
- The edit button becomes visible when hovering over any bookmark item in the displayed bookmark list.
- Clicking the edit button opens a bookmark flyout edit.
- In the flyout, users can make modifications to the bookmark's name and URL.
2. Duplicate Bookmark Prevention:
- It is not possible to create duplicate copies of bookmarks.
- Each URL can have only one bookmark associated with it.
- However, users can have multiple bookmarks with the same name.
---
### 🔍 Testing Scenarios

I have tested the following scenarios:
1. Add Bookmark:
- Verify that clicking the star icon displays a flyout to edit the bookmark's URL and name.
- Save the bookmark and confirm it is added successfully.
2. Edit Bookmark:
- Hover over a bookmark item to show the edit button.
- Click the edit button and ensure the bookmark flyout edit appears.
- Modify the URL and name, save changes, and verify the bookmark is updated.
3. Remove Bookmark:
- Hover over a bookmark item to show the edit button.
- Click the edit button and confirm the bookmark flyout edit appears.
- Remove the bookmark and ensure it is successfully deleted.
4. Bookmark Icon Indicator:
- Check that the star icon on the address bar reflects bookmark status.
- Verify the icon changes appearance (filled/unfilled) based on bookmarked status.
- Add a bookmark and confirm the star icon reflects the bookmarked state.
5. View Bookmarks:
- Click the menu flyout button and select "Bookmarks."
- Verify the flyout displays the list of bookmarks correctly.
6. Duplicate Bookmark URLs:
- Test creating duplicate bookmarks for the same URL.
- Ensure only one bookmark is allowed per unique URL.
- Verify names can be repeated for different bookmarks.
7. Bookmark Persistence:
- Navigate through different pages and sections.
- Confirm bookmarks remain intact and accessible.
- Verify bookmarks persist even after closing and reopening the application.
8. No Bookmarks:
- Remove all bookmarks and check if the UI indicates no bookmarks.
- Verify appropriate messaging when no bookmarks are present.
---
### 🖼️ Screenshots
![Screenshot 2023-06-10 at 1 46 11 PM](https://github.com/responsively-org/responsively-app/assets/59393936/d18b573c-2204-406d-a0b7-92ca94156c77)
![Screenshot 2023-06-10 at 1 45 33 PM](https://github.com/responsively-org/responsively-app/assets/59393936/62e9499c-aa17-404b-8cb9-3e37322cf6be)
![Screenshot 2023-06-10 at 1 45 45 PM](https://github.com/responsively-org/responsively-app/assets/59393936/12fc714d-0516-42a8-ae7e-ce5d62143598)
![Screenshot 2023-06-10 at 1 46 41 PM](https://github.com/responsively-org/responsively-app/assets/59393936/d7332397-9b59-4b90-a400-3ba83857dd98)
![Screenshot 2023-06-10 at 1 47 03 PM](https://github.com/responsively-org/responsively-app/assets/59393936/6e319c5a-b10c-4f6e-bfa4-27de25d08c3f)
